concordium-consensus/src/Concordium/GlobalState/Persistent/Accounts.hs

```diff
@@ -168,35 +285,71 @@ getAccountByCredId cid accs@Accounts{..} =
         Nothing -> return Nothing
         Just ai -> fmap (ai,) <$> indexedAccount ai accs
 
--- | Get the account at a given index (if any).
-getAccountIndex :: (IsProtocolVersion pv, MonadBlobStore m) => AccountAddress -> Accounts pv -> m (Maybe AccountIndex)
-getAccountIndex addr Accounts{..} = AccountMap.lookup addr accountMap
+-- | Get the 'AccountIndex' for the provided 'AccountAddress' (if any).
+--  First try lookup in the in-memory difference map associated with the the provided 'Accounts pv',
+--  if no account could be looked up, then we fall back to the lmdb backed account map.
+--
+-- If account alises are supported then the equivalence class 'AccountAddressEq' is used for determining
+-- whether the provided @AccountAddress@ is in the map, otherwise we check for exactness.
+getAccountIndex :: forall pv m. (SupportsPersistentAccount pv m) => AccountAddress -> Accounts pv -> m (Maybe AccountIndex)
+getAccountIndex addr Accounts{..} = do
+    mAccountDiffMap <- liftIO $ readIORef accountDiffMapRef
+    case mAccountDiffMap of
+        Absent -> lookupDisk 0
+        Present accDiffMap ->
+            if supportsAccountAliases (protocolVersion @pv)
+                then
+                    DiffMap.lookupViaEquivalenceClass (accountAddressEmbed addr) accDiffMap >>= \case
+                        Just accIdx -> return $ Just accIdx
+                        Nothing -> do
+                            diffMapSize <- length <$> DiffMap.flatten accDiffMap
+                            lookupDisk $ fromIntegral diffMapSize
+                else
+                    DiffMap.lookupExact addr accDiffMap >>= \case
+                        Just accIdx -> return $ Just accIdx
+                        Nothing -> do
+                            diffMapSize <- length <$> DiffMap.flatten accDiffMap
```

Comment:
Is there no better way of keeping track of the size?

In normal operation this should be quick enough, since few accounts are created per block, and there are few non-finalized ones. But it's a bit unsatisfactory.

Reply:
We can cache it per layer.